### PR TITLE
Rebuild on mid build file modification

### LINKED ISSUE #2426: cargo build does not rebuild if a source file was modified during a build
I haven't checked, but I would guess this is a timestamp issue: the input is modified before the output artifact is created, even though the modified input was not used to build the artifact. This would happen any time after rustc is parsing.

### PR DESCRIPTION
**Pull request by Eh2406**
This is an attempt to Fixes #2426, it is based on the test @dwijnand made for #5417.

### CODE REVIEW

**Comment by Eh2406** [comment]
That clean up mess things up. I forced pushed to the previous one. CI is green.

**Comment by dwijnand** [comment]
LGTM, but I thought I saw @ehuss exhibit concern or at the very least ask about whether a dedicated file is necessary, on Discord, but I can't find it now and I might've got my patches mixed up.  Looking for a second opinion 😄

**Comment by Eh2406** [comment]
The concern was [raised](https://github.com/rust-lang/cargo/pull/6477#issuecomment-449657590) on my other PR adding a timestamp file. 

For this PR I do not see a way to do it without a new file. (Assuming we want to gen the mtime from a file, as described [hear](https://github.com/rust-lang/cargo/issues/2426#issuecomment-395950178), ruffly that the file system may be on a different clock then `Systomtime::new()`.) At the start of the compile we do not yet have the info we need to write any of the existing files, so we need to put filler content into whichever file we make. After the build, if we override the the file then we will lose the mtime we care about.

Now that I am thinking about it, this may now have a new race condition.

1. Make input files at t = 0
2. Start build, making a timestamp at t = 1 
3. Rustc terminated without writing an output file.
4. Start a new build, the input files all have a mtime less then the timestamp, so no rebuild is needed.

The fix is eazy we use the `min(timestamp.mtime, output.mtime)`, but I don't know how to write a test.

File: tests/testsuite/build.rs [hunk 1]
```diff
@@ -2384,6 +2386,112 @@ fn rebuild_preserves_out_dir() {
         .run();
 }
 
+#[test]
+fn rebuild_on_mid_build_file_modification() {
```
Comment (ehuss) [inline tests/testsuite/build.rs]: I would put `rebuild_on_mid_build_file_modification` in `freshness.rs`. That is where most of the fingerprint tests are located, and `build.rs` is crazy long.

**Comment by ehuss** [comment]
> Now that I am thinking about it, this may now have a new problem.

I don't think I understand your concern. It will still rebuild the target because the new, computed fingerprint's hash does not match the original fingerprint hash. And the fingerprint isn't saved if it fails. I think the main consequences is that the log output won't be correct (the "stale" entries, and maybe the `compare()` mtime checking).

File: src/cargo/core/compiler/fingerprint.rs [hunk 2]
```diff
@@ -241,7 +242,10 @@ impl Fingerprint {
             match *local {
                 LocalFingerprint::MtimeBased(ref slot, ref path) => {
                     let path = root.join(path);
-                    let mtime = paths::mtime(&path)?;
+                    let mtime = min(
```
[the remark this answers is not in the document]
Comment (Eh2406) [reply src/cargo/core/compiler/fingerprint.rs]: I will back it out. Maybe that is why I could not make a test for it. :-P

File: tests/testsuite/build.rs (hunk 1 above)
Comment (Eh2406) [reply tests/testsuite/build.rs]: I will move it.

**Comment by Eh2406** [comment]
I moved the test, and removed the `min`.
Does that look better?

**Comment by ehuss** [comment]
I've been kicking this around, and I can't find any problems. However, the fingerprint stuff is pretty subtle so I'd like it if Alex could take a look before merging.

**Comment by alexcrichton** [comment]
Thanks for this @Eh2406! I've only got a few concerns about this:

* I think the `foo.with_file_name("invoked.timestamp")` pattern may want to be refactored a bit to reduce duplication throughout. I was also initially worried that with the same file name this would cause a lot of targets to collide onto the same file, but I think these files we're making are all in the `.fingerprint` folder in unique directory names already? I seem to recall though that multiple targets for one package go into the same directory...

* I'm a little worried about the implications for this on low-resolution timestamp filesystems. I think this means that if your filesystem has a timestamp resolution of N time units then if a build script or rustc executes in less than N time units you'll basically always be rebuilding, right? The previous behavior I believe is that you'd get one spurious recompile but it'd be stable after that, but this may be bad that it keeps rebuilding.

**Comment by Eh2406** [comment]
> refactored a bit to reduce duplication throughout.

Good point. I was trying to limit the state that needed to be passed around. Do you think a helper function would look better? Or, should I add a `timestamp_path` arg to keep track of it? Or, is there something else you had in mind?

> `.fingerprint` folder in unique directory names already?

I think the [build-script](https://github.com/rust-lang/cargo/pull/6484/files#diff-36216dee2e5190f6c670787ea469e00aR336) one may be in the `build` folder, but still in a directory with metadata hash in the name.

> I seem to recall though that multiple targets for one package go into the same directory...

Interesting! I am new to this part of the code, so I can easily miss this kind of corner case. Do you have a test you can point me to, or some way to reproduce this behavior?

> I'm a little worried about the implications for this on low-resolution timestamp filesystems.

Me to. I am hoping that someone with access to one can test whether we get this correct. That being said I don't see how this ends up infinitely looping.

- 1.1 file modified with a mtime of `1`
- 1.2 build started, making a timestamp with a mtime of `1`
- 1.3 file modified with a mtime of `1` (this may or may not happen, the important part is that the code has to assume it did.)
- 2.1 a new build is started, it sees that the mtime of the file == the mtime of the timestamp so it has to assume that 1.3 happened and starts compiling, making a timestamp with a mtime of `2`
- 3.1 a new build is started, it sees that the mtime of the file < the mtime of the timestamp so it uses the existing artifacts.

**Comment by ehuss** [comment]
I ran some tests on HFS (1 second resolution) on nightly vs this PR. The test is `touch src/lib.rs && cargo build && cargo build && ...`, and counting how many times `cargo build` is run before it considers everything fresh. For medium/large projects (like cargo), I didn't see any difference — it always builds just once as expected. Only for small projects (like bitflags or "hello world") does it run fast enough to rebuild multiple times. With an empty lib project, here's the comparison:

Tries | 6484 | nightly
------|------|--------
1     | 3    | 0 
2     | 0    | 1
3     | 25   | 9
4     | 65   | 44
5     | 629  | 748
6     | 258  | 191
7     | 20   | 7

You can read this chart as on nightly, 748 out of 1000 attempts `cargo build` ran 5 times before it was considered fresh (~75% of the time).  They numbers are roughly similar, although this PR has a wider spread/variance. I'd like to dig a little more to better understand why.

**Comment by Eh2406** [comment]
Note that if we decided that the definition of `touch` being considered in https://github.com/rust-lang/cargo/pull/6477#issuecomment-450963800 is adequate then there is a much simpler impl for this PR. The reason I am suspicious of that straightforward impl, is @matklad's comment at <https://github.com/rust-lang/cargo/issues/2426#issuecomment-395950178>. Am I just being paranoid, and it safe to `mtime = SystemTime::now()`?

**Comment by Eh2406** [comment]
Inspired by my last comment, I just made a commit that modifies the mtime of the output files to match the timestamp file. This means that the "invoked.timestamp" file is now a temp file needed for only a few cycles, to determine the mtime, and questions if it will be overwritten by other invocation no longer matter. It can be removed entirely if we decided `SystemTime::now()` is good enough.

**Comment by alexcrichton** [comment]
@ehuss if you change [this block](https://github.com/rust-lang/cargo/blob/21fe000133d8f3422f372991ac7a7cdd08f5313e/src/cargo/core/compiler/fingerprint.rs#L764-L785) to use `>` instead of `>=`, does it move 100% of tries on your script to just one invocation? (that is what I suspect is the culprit here)

@Eh2406 the current PR looks pretty good to me, I think I like the idea (if perf is acceptable which given the last time I saw your laptop I assume it is because you're building this) of using the filesystem clock as the source of truth. Perhaps a utility function could be added like `get_current_filesystem_time(dir: &Path) -> io::Result<SystemTime>`?

**Comment by Eh2406** [comment]
`get_current_filesystem_time` added. We can't switch to `>`, as the file may have been modified after we started the build, but in the same time slot (my 1.3 above).

**Comment by alexcrichton** [comment]
Oh right for sure we can't change to `>` for correctness reasons, but I wanted to confirm/deny whether that's why we spuriously consider things out of date

**Comment by ehuss** [comment]
Oh, yea, it's definitely the `>=` thing. I just wanted to better understand how much worse this makes it.  I've been doing some more testing, and it is a bit worse.  Using `toml-rs` which I consider a smallish project, doing `touch src/lib.rs && cargo build && cargo build` and checking if the 2nd build is fresh, with this PR it recompiles 72% of the time on my fast machine, whereas on nightly it is only 1%.

And just to be clear, the issue is if "time to launch cargo and start compiling" is under 1 second, then it can cause a recompile, whereas currently it is "time to launch cargo, and *finish* compiling".

I don't think it's a showstopper, just something to be concerned about.  I don't think it will be a massive problem.  CI like Travis with MacOS is probably slow enough that it is extremely unlikely the time from checkout to first build is under a second.  It might be more of an issue for people who run `cargo check` automatically on save in their editor.

I don't see any workaround other than sandboxing or hashing the contents — which I looked into before and looks tricky because we don't know what to hash until *after* the first compilation is finished.

**Comment by alexcrichton** [comment]
Hm ok no I feel like I'm seeing this problem more and more with fast machines and fast-compiling crates. (on today's nightly). We also discovered that although Linux has nanosecond resolution time stamps it doesn't actually update every nanosecond, but rather the filesystem clock updates at a coarser granularity (like every 100-200 ms or something like that). That makes me even more worried about causing spurious recompiles with this PR. 

To make matters worse, the fastest compiling crates are often at the bottom of the dependency graph, which could cause very large recompilations!

I think that we probably need to fix this `>=` issue, but via a different route. I think `>=` is the morally right condition to have here but the wrong implementation. I think that we need to do something along the lines of tracking mtimes but not relying on the filesystem to store the actual mtime itself, but rather we should do something like serialize an mtime into a file. For example this PR updates the mtime on various files but what if we did the following:

* When rustc/build scripts start, we create a file and learn the mtime from the filesystem about that file.
* When the job ends, we serialize the mtime, plus one nanosecond, into the file. 
* Later, Cargo deserializes the file to learn the mtime, reading the contents of the file rather than the mtime of the file

Something like that I think handles the case of coarse filesystem clocks, but it may not handle the case where a file is modified during that second and the filesystem doesn't realize it. 

I'm curious what others think of this though?

**Comment by Eh2406** [comment]
> To make matters worse, the fastest compiling crates are often at the bottom of the dependency graph, which could cause very large recompilations!

Can you elaborate on the use case of building a large project with a deep dependency that is a file based project? (if it is not a file, then I don't think we look at the mtime.) Is this testing a `[patch]` section on a big project? Is this a common case do to the oddities of the build system for servo/rust?

> Later, Cargo deserializes the file to learn the mtime, reading the contents of the file rather than the mtime of the file

And compares it to what? The mtime of the input files? So that makes `mtime` much more precise but we are still comparing it to the imprecise [`mtime2`](https://github.com/rust-lang/cargo/blob/9bfaf2ec09837581038967c75c33c610205536c6/src/cargo/core/compiler/fingerprint.rs#L780), so that comparison is still going to be imprecise.

**Comment by alexcrichton** [comment]
@Eh2406 ah yeah that's a good point, but I mean a file (or `[patch]`) dependency deep down in the project. I've seen a lot of spurious recompiles in rust-lang/rust recently when developing locally, and I think it's due to this `>=` vs `>` issue. In rust-lang/rust at least almost everything is a path dependency :)

I was assuming yeah we'd compare the mtime against the actual filesystem mtime. Agreed it doesn't solve the problem :(

I don't really know what to do here. On one hand the current state of affairs is pretty bad because it makes recompilation much more common. On the other hand reverting `>=` and `>` is bad because it means that in corner cases you can not rebuild when you actually need to. On filesystems with granular timestamp resolution we're just sort of fundamentally hosed because we can't tell where within a second a file was modified.

Our current state of affairs feels wrong to me though. We've introduced a pretty common bug (recompiling for fast crates) for fixing an edge case (modifying a file very quickly after a build).

**Comment by Eh2406** [comment]
Ok so it is common in rustc, that is an important user. Is it possible the spurious recompiles you are seeing are #6493?

As a user on a slow computer, the "current state of affairs" on master is pretty terrible. I often end up modifying a file in the several mins while the build is running, then I am in the "you can not rebuild when you actually need to" situation without any corner cases needed. I often end up submitting PRs that don't build at all, because `cargo test` spuriously passes locally. And I am a power user that has muscle memory to save before build.

That being said, this does make that corner case (granular timestamp + build on save) more of a problem.

**Comment by alexcrichton** [comment]
Er sorry I think we've definitely digressed a bit, the problem I'm describing of `>` vs `>=` I believe is actually orthogonal to what this PR is trying to solve. I think it's a technical limitation that our usage of `>=` makes solving this PR's issue at all much more difficult and impactful. 

In other words, I definitely agree it's worthwhile to fix this PR's issue, and I'd personally be comfortable with the current PR *if* we switched to `>` instead of `>=` when comparing mtimes.

I think at this point the only way to solve everything is cryptographic hashes. I think we could develop a system where if the mtimes were exactly the same we compared hashes of before/after, but otherwise we continue to use mtimes as cheap checks that'd be plausible.

**Comment by Eh2406** [comment]
I switched to `>` from `>=`, if this merges then #5918 should be reopened, and we should move the discussion of how to fix it there.

**Comment by alexcrichton** [comment]
I've [written up a comment](https://github.com/rust-lang/cargo/issues/5918#issuecomment-451587876) about how I see a way forward solving both this and the original issue with `>=`

**Comment by Eh2406** [comment]
I don't have access to an HFS system. I can iteratively fix theas using CI, but it will take a while.
A thought I have been having, could the filetime crate have a environment variable then wen set and in debug, always rounds times to the second, so we can test this on non limited systems?

Meanwhile, how should we proceed:

a. we continue with `>` ether by me fixing on CI or someone else offering to fix.
b. we reject this until an implementation of a hashing solution.
c. we merge this with `>=` and priorities a hashing solution.

Thoughts?

**Comment by Eh2406** [comment]
Someone who knows this stuff better than I, did this also fix the corresponding case with `cargo check` or does it use a different code path?

**Comment by Eh2406** [comment]
I tryed changing 'build' to 'check' in the test, and it failed. So I think something must be different.

**Comment by Eh2406** [comment]
turned out to be a problem with the test, the test was looking for the word compiling not the word checking.

**Pull request merged**